docs/broker/filtering.md

```diff
@@ -0,0 +1,360 @@
+deprecation.
+
+For example, a javascript filter could be added without disrupting existing CEL
+filters:
+
+```yaml
+spec:
+  filter:
+    javascript:
+      expression: ce.type == "com.github.pull.create"
+```
+
+## Alternatives Considered
+
+The following alternative solutions were considered but rejected.
+
+### Javascript or Lua as expression language
+
+Trigger filtering is a very limited case for an embedded language. All that is
+needed from a language is the ability to write conditions that evaluate to true
+or false.
+
+Javascript and Lua are commonly used as embedded languages, but their
+capabilities extend well beyond the needs of Trigger filtering. Those
+capabilities come with complexity: Javascript and Lua programs may not provably
+terminate, may have memory leaks, and may consume excessive amounts of CPU. All
+these cases would need to be guarded against by any process involved in
+filtering. If the expression language is used for upstream filter propagation,
+this could be a large number of diverse processes.
+
+Due to this additional implementation complexity, both Javascript and Lua have
+been judged less suitable for Trigger filtering than CEL.
+
+### Rego (OPA)
+
+[Rego](https://www.openpolicyagent.org/docs/v0.10.7/language-reference/) is
+already as a policy expression language by
+[Open Policy Agent](https://www.openpolicyagent.org/). It could be an attractive
+choice, but
+[OPA has indicated a desire to replace Rego with CEL](https://github.com/open-policy-agent/opa/issues/811#issuecomment-401844999).
+Since Rego doesn't seem to be used outside OPA, choosing Rego for Trigger
+filters would likely make Triggers the only remaining use of Rego.
+
+The possible maintenance burden of being the only Rego user makes Rego less
+suitable for Trigger filtering than CEL.
+
+### Custom expression language
+
+We could develop our own custom expression language for Trigger filters. This
+might be attractive since we could design the language for the needs of Trigger
+filters specifically.
+
+A possible example is Kubernetes' set-based label selectors:
+
+```
+environment in (production, qa)
+tier notin (frontend, backend)
+partition
+!partition
+```
+
+The disadvantage of this approach is that the Knative Eventing contributors
+would be responsible for implementing, documenting, and maintaining the
+language. CEL already meets the current needs of Trigger filters and is
+extensible to support future needs.
+
+The implementation and maintenance work required to develop a custom expression
+language make this solution less suitable for Trigger filtering than CEL.
+
+### Structured filters
+
+Instead of using an expression language, we could design a system of structured
+filters expressible as yaml or JSON, similar to Kubernetes set-based label
+selectors:
+
+```yaml
+matchExpressions:
+  - {key: tier, operator: In, values: [cache]}
+  - {key: environment, operator: NotIn, values: [dev]}
+```
```

Comment:
This is good motivation for curating simple data-driven experiences since they're easier to support in a variety of platforms with or without CEL as the evaluation engine under the covers. The expression should be an escape hatch and may not be supported on all platforms.